src/core/qgsexpressioncontext.cpp

```diff
@@ -769,3 +769,45 @@ QgsFeedback *QgsExpressionContext::feedback() const
 {
   return mFeedback;
 }
+
+QString QgsExpressionContext::uniqueHash( bool &ok, const QSet<QString> &variables ) const
+{
+  QString hash;
+  ok = true;
+  const QString delimiter( QStringLiteral( "||~~||" ) );
+
+  if ( hasFeature() )
+  {
+    hash.append( QString::number( feature().id() ) + delimiter + QString::number( qHash( feature() ) ) + delimiter );
+  }
+
+  QStringList sortedVars = qgis::setToList( variables );
+  if ( sortedVars.empty() )
+    sortedVars = variableNames();
+  std::sort( sortedVars.begin(), sortedVars.end() );
+
+  for ( const QString &variableName : std::as_const( sortedVars ) )
+  {
+    const QVariant value = variable( variableName );
+    if ( QgsVariantUtils::isNull( value ) )
+    {
+      hash.append( delimiter );
```

Comment:
Shall we not include `variableName in the hash? 2 different variable with the same value could have the same hash (with a different set of given variables but still)

Reply:
Yes, good point! I'll fix